src/core/thread/network_data_leader_ftd.cpp

```diff
@@ -152,6 +152,8 @@ void Leader::HandleServerData(Coap::Message &aMessage, const Ip6::MessageInfo &a
 
     otLogInfoNetData("Received network data registration");
 
+    VerifyOrExit(aMessageInfo.GetPeerAddr().IsRoutingLocator());
```

Comment:
One concern is that if putting the check here, in case there is SRV_DATA.ntf with non RLOC source, there would be no ACK for the CON Coap message, thus there might be meaningless CoAP retransmission.
In case there is some coincidence which you cares, maybe move the check before trying to retrieve the short address from source?

Reply:
There are two concerns regarding this PR:
1. Should we check before `RegisterNetworkData`?
2. Should we send empty ACK?

For 1), I think we should not check before `RegisterNetworkData` because the caller would thought that the request is fully processed however it is not. 

For 2), I think we should not reply with `ACK 2.04 Changed`, but reply with `ACK 4.00 Bad Request`. However current OT codebase rarely does that for invalid COAP requests, but only exit skip quietly. 

Thoughts?

Reply:
> There are two concerns regarding this PR:
> 
> 1. Should we check before `RegisterNetworkData`?
> 2. Should we send empty ACK?
> 
> For 1), I think we should not check before `RegisterNetworkData` because the caller would thought that the request is fully processed however it is not.

I think the current location is fine. As noted, sending with non-RLOC IPv6 Source Address is unexpected behavior and we should catch that as early as possible.

> For 2), I think we should not reply with `ACK 2.04 Changed`, but reply with `ACK 4.00 Bad Request`. However current OT codebase rarely does that for invalid COAP requests, but only skip quietly.

Yes, the existing codebase generally does not respond with Bad Request for TMF messages. We can consider changing that in a separate PR if we find that useful.

Reply:
I think always sending ACK for CON is helpful to avoid meaningless CoAP retransmission as @librasungirl said.

Reply:
It makes sense to catch it earlier if invalid.

Generally SRV_DATA.ntf in Thread-conformance implementation should have the right RLOC source. And the issue we observed so far only happens in the special test scenario which constructs fake SRV_DATA.ntf and sends it over the wpantund which filters out RLOC address. we may keep it here and change to support Bad Request to avoid the impact of CoAP retransmission when required in the future? @simonlingoogle

Reply:
> we may keep it here and change to support Bad Request to avoid the impact of CoAP retransmission when required in the future? @simonlingoogle

Agree.